slack_bolt/workflows/step/step.py

```diff
@@ -1,19 +1,263 @@
+        lazy: Optional[List[Callable[..., None]]] = None,
+    ):
+        """Register a new edit listener with details.
+        You can use this method as decorator as well.
+
+        @my_step.edit
+        def edit_my_step(ack, configure):
+            pass
+
+        It's also possible to add additional listener matchers and/or middleware
+
+        @my_step.edit(matchers=[is_valid], middleware=[update_context])
+        def edit_my_step(ack, configure):
+            pass
+        """
+
+        if len(args) == 1:
+            func = args[0]
+            self._edit = self._to_listener("edit", func, matchers, middleware)
+            return func
+
+        def _inner(func):
+            functions = [func] + (lazy if lazy is not None else [])
+            self._edit = self._to_listener("edit", functions, matchers, middleware)
+
+            @wraps(func)
+            def _wrapper(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            return _wrapper
+
+        return _inner
+
+    def save(
+        self,
+        *args,
+        matchers: Optional[Union[Callable[..., bool], ListenerMatcher]] = None,
+        middleware: Optional[Union[Callable, Middleware]] = None,
+        lazy: Optional[List[Callable[..., None]]] = None,
+    ):
+        """Register a new save listener with details.
+        You can use this method as decorator as well.
+
+        @my_step.save
+        def save_my_step(ack, step, update):
+            pass
+
+        It's also possible to add additional listener matchers and/or middleware
+
+        @my_step.save(matchers=[is_valid], middleware=[update_context])
+        def save_my_step(ack, step, update):
+            pass
+        """
+
+        if len(args) == 1:
+            func = args[0]
+            self._save = self._to_listener("save", func, matchers, middleware)
+            return func
+
+        def _inner(func):
+            functions = [func] + (lazy if lazy is not None else [])
+            self._save = self._to_listener("save", functions, matchers, middleware)
+
+            @wraps(func)
+            def _wrapper(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            return _wrapper
+
+        return _inner
+
+    def execute(
+        self,
+        *args,
+        matchers: Optional[Union[Callable[..., bool], ListenerMatcher]] = None,
+        middleware: Optional[Union[Callable, Middleware]] = None,
+        lazy: Optional[List[Callable[..., None]]] = None,
+    ):
+        """Register a new execute listener with details.
+        You can use this method as decorator as well.
+
+        @my_step.execute
+        def execute_my_step(step, complete, fail):
+            pass
+
+        It's also possible to add additional listener matchers and/or middleware
+
+        @my_step.save(matchers=[is_valid], middleware=[update_context])
+        def execute_my_step(step, complete, fail):
+            pass
+        """
+
+        if len(args) == 1:
```

Comment:
What do you think about delegating this conditional to a method named something like `used_without_argument` or maybe `used_as_plain_decorator`?

Eventuals refactors to this logic may be easier by changing it in only one place, we would need to pass the type of listener to that method

Reply:
It sounds nice. I will have some method to commonize the logic. Thanks for the feedback!